[scroll-animations] Have ScrollTimeline hold a Styleable source

nmoucht opened the pull request:
#### 5ab04d6e08be21da84cf0cdc63115faa7702e99d
<pre>
[scroll-animations] Have ScrollTimeline hold a Styleable source
<a href="https://bugs.webkit.org/show_bug.cgi?id=286504">https://bugs.webkit.org/show_bug.cgi?id=286504</a>
<a href="https://rdar.apple.com/143590723">rdar://143590723</a>

Reviewed by Tim Nguyen.

To handle ScrollTimelines associated with pseudo elements, have ScrollTimeline hold a
Styleable source. This will need a follow up for ViewTimeline code as well as the named
timeline code on AnimationTimelinesController.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/pseudo-on-scroller-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):
(WebCore::ScrollTimeline::setSource):
(WebCore::ScrollTimeline::controller const):
(WebCore::ScrollTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::ScrollTimeline::animationTimingDidChange):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::applyKeyframeEffects const):
* Source/WebCore/style/Styleable.h:
(WebCore::WeakStyleable::operator== const):
(WebCore::WeakStyleable::WeakStyleable):
(WebCore::WeakStyleable::element const):
(WebCore::WeakStyleable::pseudoElementIdentifier const):
(WebCore::WeakStyleable::updateElement):
(WebCore::Styleable::applyKeyframeEffects const): Deleted.

Canonical link: <a href="https://commits.webkit.org/289523@main">https://commits.webkit.org/289523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50ab1c2979ce1eeae7cb7c83888a30a90631074c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92001 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37879 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89190 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6926 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67350 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25088 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90142 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78872 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47670 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5072 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33250 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36996 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75560 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93886 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10400 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76154 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14506 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74728 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75355 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19691 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18129 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7219 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13588 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14321 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19614 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17509 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15847 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/Styleable.h:
```diff
@@ -204,20 +200,34 @@ class WeakStyleable {
+    bool operator==(const WeakStyleable& other) const
+    {
+        return (element() == other.element() && pseudoElementIdentifier() == other.pseudoElementIdentifier());
+    }
```
Is this not just the default operator?

nmoucht replied:
Yes.

nt1m commented on Source/WebCore/style/Styleable.h:
```diff
@@ -204,20 +200,34 @@ class WeakStyleable {
         return *this;
     }
 
+    WeakStyleable(const Styleable& styleable)
+    {
+        m_element = styleable.element;
+        m_pseudoElementIdentifier = styleable.pseudoElementIdentifier;
+    }
```
I'm curious what in this patch uses this.

nmoucht replied:
No longer necessary.

nt1m replied:
Can we remove it if that's the case?

nmoucht replied:
Apologies this is necessary for the `setSource(Styleable::fromElement(*source))` call.

graouts commented on Source/WebCore/animation/CSSAnimation.cpp:
```diff
@@ -160,7 +160,10 @@ void CSSAnimation::syncStyleOriginatedTimeline()
         }, [&] (const Animation::AnonymousScrollTimeline& anonymousScrollTimeline) {
             auto scrollTimeline = ScrollTimeline::create(anonymousScrollTimeline.scroller, anonymousScrollTimeline.axis);
-            scrollTimeline->setSource(target.ptr());
+            auto weakStyleable = WeakStyleable();
```
```suggestion
            WeakStyleable weakStyleable();
```

nmoucht replied:
Made this use Styleable.

graouts commented on Source/WebCore/animation/KeyframeEffect.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "AcceleratedEffect.h"
+#include "Animation.h"
```
Why is this change necessary?

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts commented on Source/WebCore/Headers.cmake:
```diff
@@ -1137,6 +1137,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     dom/Position.h
     dom/ProcessingInstruction.h
     dom/ProgressEvent.h
+    dom/PseudoElement.h
```
Why is this necessary?

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -29,11 +29,13 @@
 #include "CSSAnimation.h"
 #include "CSSNumericFactory.h"
 #include "CSSNumericValue.h"
+#include "CSSParserContext.h"
```
Not clear why this is relevant to this patch. If this addressing a unified build issue, perhaps it would be better to address this first as a standalone commit. I get the feeling this patch contains several of those. At the very least, these should be called out in the ChangeLog.

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -43,8 +44,6 @@ class ScrollableArea;
 struct TimelineRange;
 
-enum class Scroller : uint8_t { Nearest, Root, Self };
-
```
Why is that change necessary?

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -54,7 +53,8 @@ class ScrollTimeline : public AnimationTimeline {
     virtual Element* source() const;
-    void setSource(const Element*);
+    void setSource(Element*);
+    void setSource(const WeakStyleable&);
```
I don't think this is a good idea. We should only use a `Styleable` just like we do throughout the animations code.

nmoucht replied:
Made this use Styleable.

graouts commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -42,12 +42,6 @@ class Element;
-    std::optional<Length> start;
-    std::optional<Length> end;
-    bool operator==(const ViewTimelineInsets&) const = default;
-};
```
Why is that change being made?

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -294,6 +294,11 @@ void Styleable::willChangeRenderer() const
+OptionSet<AnimationImpact> Styleable::applyKeyframeEffects(RenderStyle& targetStyle, UncheckedKeyHashSet<AnimatableCSSProperty>& affectedProperties, const RenderStyle* previousLastStyleChangeEventStyle, const Style::ResolutionContext& resolutionContext) const
+{
+    return element.ensureKeyframeEffectStack(pseudoElementIdentifier).applyKeyframeEffects(targetStyle, affectedProperties, previousLastStyleChangeEventStyle, resolutionContext);
+}
```
Why is this change necessary?

nmoucht replied:
Moved build fixes to https://github.com/WebKit/WebKit/pull/39640.

graouts reviewed (changes requested):
I think this PR shouldn't land as-is, it seems to mix up some refactoring / unified fixes with the relevant change to fix the bug.

nmoucht committed:
[scroll-animations] Have ScrollTimeline hold a Styleable source
https://bugs.webkit.org/show_bug.cgi?id=286504
rdar://143590723

Reviewed by Tim Nguyen.

To handle ScrollTimelines associated with pseudo elements, have ScrollTimeline hold a
Styleable source. This will need a follow up for ViewTimeline code as well as the named
timeline code on AnimationTimelinesController.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/pseudo-on-scroller-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):
(WebCore::ScrollTimeline::setSource):
(WebCore::ScrollTimeline::controller const):
(WebCore::ScrollTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::ScrollTimeline::animationTimingDidChange):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::applyKeyframeEffects const):
* Source/WebCore/style/Styleable.h:
(WebCore::WeakStyleable::operator== const):
(WebCore::WeakStyleable::WeakStyleable):
(WebCore::WeakStyleable::element const):
(WebCore::WeakStyleable::pseudoElementIdentifier const):
(WebCore::WeakStyleable::updateElement):
(WebCore::Styleable::applyKeyframeEffects const): Deleted.

Canonical link: https://commits.webkit.org/289523@main
